src/OSPSuite.SimModel/src/TableFormulaWithXArgument.cpp

```diff
@@ -0,0 +1,235 @@
+{
+	const char * ERROR_SOURCE = "TableFormulaWithXArgument::DE_Jacobian";
+
+	assert ((_tableObject != NULL) && (_XArgumentObject != NULL));
+
+	if (preFactor == 0.0)
+		return;
+
+	bool forCurrentRunOnly = true;
+	if (_XArgumentObject->IsConstant(forCurrentRunOnly))
+		return;
+
+	throw ErrorData(ErrorData::ED_ERROR, ERROR_SOURCE, "Table formula with X argument does not support time dependent argument objects");
+}
+
+Formula* TableFormulaWithXArgument::DE_Jacobian(const int iEquation)
+{
+	return _tableObject->DE_Jacobian(iEquation); // TODO: fix
+}
+
+Formula * TableFormulaWithXArgument::clone()
+{
+	return new ConstantFormula(0.0);
+	TableFormulaWithXArgument * f = new TableFormulaWithXArgument();
+
+	// TODO: fix
+
+	return f;
+}
+
+Formula * TableFormulaWithXArgument::RecursiveSimplify()
+{
+	return this;
+}
+
+void TableFormulaWithXArgument::SetQuantityReference (const QuantityReference & quantityReference)
+{
+	assert ((_tableObject != NULL) && (_XArgumentObject != NULL));
+}
+
+bool TableFormulaWithXArgument::IsZero(void)
+{
+	assert ((_tableObject != NULL) && (_XArgumentObject != NULL));
+
+	const bool forCurrentRunOnly = false;
+
+	//case 1: table formula is constant (all y values are the same)
+	//        If so: calculate table formula value for ANY argument (e.g. zero) and check if it returns 0
+	if (_tableObject->IsConstant(forCurrentRunOnly))
+		return (_tableObject->GetValue(NULL, 0.0, USE_SCALEFACTOR) == 0);
+
+	//case 2: X argument is constant. Check if table formula for this const argument value is zero
+	if (_XArgumentObject->IsConstant(forCurrentRunOnly))
+	{
+		double argument = _XArgumentObject->GetValue(NULL, 0.0, USE_SCALEFACTOR);
+		return (_tableObject->GetValue(NULL, argument, USE_SCALEFACTOR) == 0);
+	}
+
+	//neither x argument nor table values are constant => formula is not const zero
+	return false;
+}
+
+bool TableFormulaWithXArgument::IsRefIndependent(double & value)
+{
+	return false;
+}
+
+vector < HierarchicalFormulaObject * > TableFormulaWithXArgument::GetUsedHierarchicalFormulaObjects()
+{
+	assert ((_tableObject != NULL) && (_XArgumentObject != NULL));
+
+	vector<HierarchicalFormulaObject * > hfos;
+
+	hfos.push_back((HierarchicalFormulaObject *)_tableObject);
+	hfos.push_back((HierarchicalFormulaObject *)_XArgumentObject);
+
+	return hfos;
+}
+
+void TableFormulaWithXArgument::Finalize()
+{
+	assert ((_tableObject != NULL) && (_XArgumentObject != NULL));
+}
+
+vector <double> TableFormulaWithXArgument::RestartTimePoints()
+{
+	assert((_tableObject != NULL) && (_XArgumentObject != NULL));
+
+	vector <double> restartTimes;
+	
+	const bool forCurrentRunOnly = false;
+	if (_XArgumentObject->IsConstant(forCurrentRunOnly))
+		return restartTimes; //argument const=>formula will always be const=>no need to restart
+
+	//TODO fix non-constant argument
+	return restartTimes;
```

Comment:
for time dependent XArgument this will become really tricky. for pH-solubility not relevant

Reply:
but Xargument = f(Time) would work? Or not even that?